uniffi_macros/src/enum_.rs

```diff
@@ -0,0 +1,144 @@
+    util::{assert_type_eq, create_metadata_static_var, try_read_field},
+};
+
+pub fn expand_enum(input: DeriveInput, module_path: Vec<String>) -> TokenStream {
+    let variants = match input.data {
+        Data::Enum(e) => Some(e.variants),
+        _ => None,
+    };
+
+    let ident = &input.ident;
+
+    let (write_impl, try_read_impl) = match &variants {
+        Some(variants) => {
+            let write_match_arms = variants.iter().enumerate().map(|(i, v)| {
+                let v_ident = &v.ident;
+                let fields = v.fields.iter().map(|f| &f.ident);
+                let idx = Index::from(i);
+                let write_fields = v.fields.iter().map(write_field);
+
+                quote! {
+                    Self::#v_ident { #(#fields),* } => {
+                        ::uniffi::deps::bytes::BufMut::put_i32(buf, #idx);
+                        #(#write_fields)*
+                    }
+                }
+            });
+            let write_impl = quote! {
+                match obj { #(#write_match_arms)* }
+            };
+
+            let try_read_match_arms = variants.iter().enumerate().map(|(i, v)| {
+                let idx = Index::from(i);
+                let v_ident = &v.ident;
+                let try_read_fields = v.fields.iter().map(try_read_field);
+
+                quote! {
+                    #idx => Self::#v_ident { #(#try_read_fields)* },
+                }
+            });
+            let error_format_string = format!("Invalid {ident} enum value: {{}}");
+            let try_read_impl = quote! {
+                ::uniffi::check_remaining(buf, 4)?;
+
+                Ok(match ::uniffi::deps::bytes::Buf::get_i32(buf) {
+                    #(#try_read_match_arms)*
+                    v => ::uniffi::deps::anyhow::bail!(#error_format_string, v),
+                })
+            };
+
+            (write_impl, try_read_impl)
+        }
+        None => {
+            let unimplemented = quote! { ::std::unimplemented!() };
+            (unimplemented.clone(), unimplemented)
+        }
+    };
+
+    let meta_static_var = variants
+        .map(|variants| {
+            let name = ident.to_string();
+            let variants_res: syn::Result<_> = variants
+                .iter()
+                .map(|v| {
+                    let name = v.ident.to_string();
+                    let fields = v
+                        .fields
+                        .iter()
+                        .map(|f| {
+                            let name = f
+                                .ident
+                                .as_ref()
+                                .ok_or_else(|| {
+                                    syn::Error::new_spanned(
+                                        v,
+                                        "UniFFI only supports enum variants with named fields \
+                                         (or no fields at all)",
+                                    )
```

Comment:
With UDL we actually somewhat support enum variants with tuple-style fields, if they're defined with the `enum` style rather than the `interface` style.  I'm hoping with the proc-macros we can properly support them.

In terms of the PR, I think erroring out is great for now.

Reply:
Oh, I didn't know that. Will look into it 🙂

Reply:
[The docs](https://mozilla.github.io/uniffi-rs/udl/enumerations.html) seem to disagree with you, it's either no fields for all variants (`enum` in UDL) or fields with names (`[Enum] interface` in UDL).

Reply:
Oops, I was thinking of errors.  Those use the same UDL syntax, but because of historical reasons when a UDL `enum` is used for an error, the Rust `enum` can actually have fields.  It's really messy and one thing I hope the proc-macros can clean up.